### PR TITLE
Add voting to projects.

### PR DESCRIPTION
**Pull request by prodex**
#14 

- [x] Allow people to rate projects. Should be +1 or -1.
- [x] Add Top 100 page.

### CODE REVIEW

File: assets/app/js/vote.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    $('.vote .button').on('click', function (e) {
+        e.preventDefault();
+
+        var el = $(this);
+        var container = el.parents('.vote');
+        var endpoint = el.data('endpoint');
+        var value = el.data('value');
+
+        if (el.is('.disabled')) {
+            return false
```
Comment (samdark) [inline assets/app/js/vote.js]: ;

File: assets/app/styles/main.less [hunk 2]
```diff
@@ -297,6 +297,12 @@ a {
         }
       }
     }
+    
+    .vote {
+      margin: 0px 0px 5px 10px;
```
Comment (samdark) [inline assets/app/styles/main.less]: 0px → 0

File: config/urls.php [hunk 3]
```diff
@@ -19,8 +19,9 @@
     [
         'class' => \yii\rest\UrlRule::class,
         'controller' => ['1.0/projects' => 'api1/project'],
-        'only' => ['index', 'view', 'update'],
+        'only' => ['index', 'view', 'update', 'voting'],
```
Comment (samdark) [inline config/urls.php]: voting → vote

File: config/urls.php [hunk 4]
```diff
@@ -19,8 +19,9 @@
...
         'prefix' => 'api',
+        'extraPatterns' =>  ['PUT,PATCH {id}/voting' => 'voting'],
```
Comment (samdark) [inline config/urls.php]: voting → vote

File: models/Project.php [hunk 5]
```diff
@@ -433,4 +435,20 @@ public function getPrimaryImage()
         
         return $this->_primaryImage;
     }
+
+    /**
+     * Return vote value.
+     * 
+     * @return int
+     */
+    public function getVoteValue()
```
Comment (samdark) [inline models/Project.php]: Better to call it `getVotingResult()`

File: models/Vote.php [hunk 6]
```diff
@@ -93,4 +94,23 @@ public function getUser()
     {
         return $this->hasOne(User::className(), ['id' => 'user_id'])->inverseOf('votes');
     }
+
+    /**
+     * Return a vote for pair the project and user.
```
Comment (samdark) [inline models/Vote.php]: Returns a vote given project id and user id.

File: modules/api1/controllers/ProjectController.php [hunk 7]
```diff
@@ -62,14 +63,60 @@ public function actionUpdate($id)
...
 
     /**
+     * @param int $id
+     *
+     * @return array
+     * @throws NotFoundHttpException
+     * @throws ServerErrorHttpException
+     * @throws UnauthorizedHttpException
+     */
+    public function actionVoting($id)
```
Comment (samdark) [inline modules/api1/controllers/ProjectController.php]: voting → vote

File: modules/api1/controllers/ProjectController.php [hunk 8]
```diff
@@ -62,14 +63,60 @@ public function actionUpdate($id)
...
+    {
+        $user = $this->getCurrentUser();
+        if (!$user) {
+            throw new UnauthorizedHttpException('User should be authorized in order to manage voting.');
+        }
+        
+        $project = Project::find()
```
Comment (samdark) [inline modules/api1/controllers/ProjectController.php]: Could be `Project::findOne($id)`.

File: modules/api1/controllers/ProjectController.php [hunk 9]
```diff
@@ -62,14 +63,60 @@ public function actionUpdate($id)
...
+            ->where(['id' => $id])
+            ->limit(1)
+            ->one();
+        
+        if (!$project) {
+            throw new NotFoundHttpException("The requested project does not exist.");
+        }
+
+        $request = Yii::$app->getRequest();
```
Comment (samdark) [inline modules/api1/controllers/ProjectController.php]: `$request` is never used except the next line. Doesn't worth introducing it.

File: modules/api1/controllers/ProjectController.php [hunk 10]
```diff
@@ -62,14 +63,60 @@ public function actionUpdate($id)
...
      * @inheritdoc
      */
     protected function verbs()
     {
         return [
             'index' => ['GET', 'HEAD'],
             'view' => ['GET', 'HEAD'],
-            'update' => ['PUT', 'PATCH']
+            'update' => ['PUT', 'PATCH'],
+            'voting' => ['PUT', 'PATCH']
```
Comment (samdark) [inline modules/api1/controllers/ProjectController.php]: voting → vote

File: modules/api1/docs.md [hunk 11]
```diff
@@ -25,6 +25,7 @@ Each project contains the following fields:
   for API calls. See [Languages](#languages).
 - `thumbnail` - URL pointing to project thumbnail image.
 - `users` - users participated in the project.
+- `voteValue` - total voting to project.
```
Comment (samdark) [inline modules/api1/docs.md]: votingResult

File: modules/api1/docs.md [hunk 12]
```diff
@@ -80,6 +81,21 @@ You may pass additional parameters when querying a list:
   returned.
 - `yiiVersion` - version of the framework project built with. Either `1.0` or `1.1`.
 
+
+### Voting to project <a href="#projects-voting" id="projects-voting">#</a>
```
Comment (samdark) [inline modules/api1/docs.md]: voting *a* project

File: modules/api1/docs.md [hunk 13]
```diff
@@ -80,6 +81,21 @@ You may pass additional parameters when querying a list:
...
+
+In order to voting to project, use the following request:
```
Comment (samdark) [inline modules/api1/docs.md]: to vote a project

File: modules/api1/docs.md [hunk 14]
```diff
@@ -80,6 +81,21 @@ You may pass additional parameters when querying a list:
...
+
+> PUT [/projects/1/voting](/en/api/1.0/projects/1/voting)
```
Comment (samdark) [inline modules/api1/docs.md]: voting → vote

File: widgets/Vote.php [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }
+    
+    /**
+     * @return string
+     */
+    public function run()
+    {
+        if (Yii::$app->user->isGuest) {
+            return '';
```
Comment (samdark) [inline widgets/Vote.php]: How about displaying the control and offering the user to log in?

File: widgets/Vote.php [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        if ($vote) {
+            $buttonClass[$vote->value][] = 'disabled';
+        }
+        
+        $voteUpHtml = Html::tag(
+            'span',
+            Html::icon('thumbs-up'),
+            [
+                'data-id' => $this->project->id,
+                'data-endpoint' => Url::to(['/api1/project/voting', 'id' => $this->project->id]),
```
Comment (samdark) [inline widgets/Vote.php]: voting → vote

File: widgets/Vote.php [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+                'class' => $buttonClass[VoteModel::VALUE_UP],
+                'title' => Yii::t('vote', 'Up'),
+            ]
+        );
+
+        $voteDownHtml = Html::tag(
+            'span',
+            Html::icon('thumbs-down'),
+            [
+                'data-endpoint' => Url::to(['/api1/project/voting', 'id' => $this->project->id]),
```
Comment (samdark) [inline widgets/Vote.php]: voting → vote

**Comment by prodex** [comment]
done

**Comment by prodex** [comment]
Maybe later need to do calculation TOP-100 via cron at some interval.

File: config/params.php [hunk 18]
```diff
@@ -13,6 +13,7 @@
     ],
 
     'project.pagesize' => 9,
+    'project.countTopProjects' => 100,
```
Comment (samdark) [inline config/params.php]: maxTopProjects

File: messages/ru/project.php [hunk 19]
```diff
@@ -86,4 +86,6 @@
     '{n, plural, one{# project} other{# projects}} made with {link}' => '{n,plural,one{# проект сделан} few{# проекта сделано} many{# проектов сделано} other{# проектов сделано}} на {link}',
     'Mark as primary image' => 'Сделать главным изображением',
     'Primary image' => 'Главное изображение',
+    'Top {n} projects' => 'Top {n} проектов',
```
Comment (samdark) [inline messages/ru/project.php]: Топ?

File: views/project/topProjects.php [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+$this->title = Yii::t('project', 'Top {n} projects', ['n' => $countTopProjects]);
+?>
+
+<div class="project-index">
+    <div class="row">
+        <?= ListView::widget([
+            'dataProvider' => $dataProvider,
+            'layout' => '{items}',
+            'options' => ['class' => 'projects-flow'],
+            'emptyText' => Yii::t('project', 'Top {n} projects still empty', ['n' => $countTopProjects]),
```
Comment (samdark) [inline views/project/topProjects.php]: is still empty

File: messages/ru/project.php (hunk 19 above)
Comment (prodex) [inline messages/ru/project.php]: I thought that in Russian the expression "Top 100" is more stable than "Топ 100". Fixed.

**Comment by samdark** [comment]
https://github.com/samdark/yiipowered/commit/9cd9ec11811b18ba91d4c7ffa51a713afec2e059 Merged! Thank you very much.

**Pull request closed without merging**